File: src/scala/engine/movegen/scala/Move.scala
package scala.engine.movegen.scala

/**
  * Created by melvic on 8/5/18.
  */
sealed trait MoveType

case object Normal extends MoveType
case object Attack extends MoveType
case object DoublePawnPush extends MoveType
case object EnPassant extends MoveType

trait Move[A] {
  def source: A
  def destination: A
  def moveType: MoveType
}

/**
  * This is a frontend-friendly Move sub-class. Use this only when you
  * plan to integrate the scala.engine with the GUI.
  */
case class LocationMove(source: Location, destination: Location, moveType: MoveType) extends Move[Location]

case class BitboardMove(source: Int, destination: Int, moveType: MoveType) extends Move[Int]

File: src/java/engine/movegen/Move.java
package java.engine.movegen;

/**
 * Created by melvic on 8/4/18.
 */
public interface Move<T> {
    public enum Type { DEFAULT, PAWN_DOUBLE_PUSH }

    public T getSource();
    public T getDestination();

    public default Type getType() {
        return Type.DEFAULT;
    }

    class Builder<T> {
        private T source;

        Builder(T source) {
            this.source = source;
        }

        public Move<T> to(T destination) {
            return new Move<T>() {
                @Override
                public T getSource() {
                    return source;
                }

                @Override
                public T getDestination() {
                    return destination;
                }
            };
        }
    }

    public static <T> Builder<T> from(T source) {
        return new Builder<T>(source);
    }
}
